SmartAppointmentSystem.Business/Implementations/DoctorUserService.cs

```diff
@@ -0,0 +1,112 @@
+
+public class DoctorUserService(AppointmentContext context, ITokenService tokenService) : IDoctorUserService
+{
+    public async Task<Doctor> GetDoctorByIdAsync(Guid id, CancellationToken cancellationToken)
+    {
+        var doctor = await context.Doctors
+            .AsNoTracking()
+            .FirstOrDefaultAsync(x => x.Id == id, cancellationToken);
+        return doctor;
+    }
+    public async Task<List<Doctor>> GetAllDoctorsAsync(CancellationToken cancellationToken)
+    {
+        var doctors = await context.Doctors
+            .AsNoTracking()
+            .ToListAsync(cancellationToken);
+        return doctors;
+    }
+    public async Task<List<Doctor>> GetDoctorsWithMostAppointmentsAsync(CancellationToken cancellationToken)
+    {
+        var doctors = await context.Doctors
+            .AsNoTracking()
+            .OrderByDescending(x=>x.Appointments.Count)
+            .ToListAsync(cancellationToken);
+        return doctors;
+    }
+    public async Task<List<DoctorsRatingDTO>> GetTopRatedDoctorsAsync(CancellationToken cancellationToken)
+    {
+        var queryResult = await context.Doctors
+            .Include(d => d.Ratings)
+            .Select(d => new
+        {
+            Doctor = d,
+            AverageRating = d.Ratings.Any() ? d.Ratings.Average(r => r.Score) : 0
+        })
+            .OrderByDescending(x => x.AverageRating)
+            .ToListAsync(cancellationToken);
+
+        var result = queryResult.Select(x =>
+        {
+            var dto = new DoctorsRatingDTO();
+            dto.Map(x.Doctor); 
+            dto.AverageRating = Math.Round(x.AverageRating, 2); 
+            return dto;
+        }).ToList();
+        return result;
+    }
+    public async Task CreateDoctorAsync(DoctorUserRequestDTO requestDTO, CancellationToken cancellationToken)
+    {
+        var hashedPassword = BCrypt.Net.BCrypt.HashPassword(requestDTO.Password);
+        var doctorEntity = requestDTO.Map();
+        doctorEntity.PasswordHash = hashedPassword;
+        await context.Doctors.AddAsync(doctorEntity, cancellationToken);
+        await context.SaveChangesAsync(cancellationToken);
+    }
+    public async Task<List<Doctor>> SearchDoctorsNameAsync(string query, CancellationToken cancellationToken)
+    {
+        if (string.IsNullOrEmpty(query))
+        {
+            return await context.Doctors.ToListAsync(cancellationToken);
+        }
+        var doctorsName = await context.Doctors
+            .AsNoTracking()
+            .Where(d => EF.Functions.Like(d.Name, $"%{query}%"))
+            .ToListAsync(cancellationToken);
+        return doctorsName;
+    }
+
+    public async Task<UserResponseModel> LoginUserAsync(DoctorUserLoginRequestDTO request, CancellationToken cancellationToken)
+    {
+        if (string.IsNullOrEmpty(request.Name) || string.IsNullOrEmpty(request.Password))
+        {
+            throw new ArgumentNullException(nameof(request));
```

Comment:
These exceptions do not make sense. If name is empty string, it will throw exception and will return 500 to the client (UI). Actually it should be badrequest instead of a internal server error (500).